src/Databases/DuckDB/DatabaseDuckDB.cpp

```diff
@@ -0,0 +1,206 @@
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <Databases/DuckDB/fetchDuckDBTableStructure.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTColumnDeclaration.h>
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/Context.h>
+#include <Storages/StorageDuckDB.h>
+#include <Databases/DuckDB/DuckDBUtils.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int DUCKDB_ENGINE_ERROR;
+    extern const int UNKNOWN_TABLE;
+}
+
+DatabaseDuckDB::DatabaseDuckDB(
+        ContextPtr context_,
+        const ASTStorage * database_engine_define_,
+        bool is_attach_,
+        const String & database_path_)
+    : IDatabase("DuckDB")
+    , WithContext(context_->getGlobalContext())
+    , database_engine_define(database_engine_define_->clone())
+    , database_path(database_path_)
+    , log(&Poco::Logger::get("DatabaseDuckDB"))
+{
+    duckdb_instance = openDuckDB(database_path_, context_, !is_attach_);
+}
+
+
+bool DatabaseDuckDB::empty() const
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    return fetchTablesList().empty();
+}
+
+
+DatabaseTablesIteratorPtr DatabaseDuckDB::getTablesIterator(ContextPtr local_context, const IDatabase::FilterByNameFunction &) const
+{
+    std::lock_guard<std::mutex> lock(mutex);
+
+    Tables tables;
+    auto table_names = fetchTablesList();
+    for (const auto & table_name : table_names)
+        tables[table_name] = fetchTable(table_name, local_context, true);
+
+    return std::make_unique<DatabaseTablesSnapshotIterator>(tables, database_name);
+}
+
+
+std::unordered_set<std::string> DatabaseDuckDB::fetchTablesList() const
+{
+    if (!duckdb_instance)
+        duckdb_instance = openDuckDB(database_path, getContext(), /* throw_on_error */true);
+
+    std::unordered_set<String> tables;
+
+    std::string query = "SHOW TABLES;";
+
+    duckdb::Connection con(*duckdb_instance);
+    auto result = con.Query(query);
+
+    if (result->HasError())
+    {
+        throw Exception(ErrorCodes::DUCKDB_ENGINE_ERROR,
+                        "Cannot fetch DuckDB database tables. Error type: {}. Message: {}",
+                        result->GetErrorType(), result->GetError());
+    }
+
+    auto chunk = result->Fetch();
+
+    if (!chunk)
+        std::cout << "\n\n\n\n\nDUCKDB: CHUNK IS EMPTY\n\n\n\n\n";
+
+    while (chunk)
+    {
+        for (size_t idx = 0; idx < chunk->size(); ++idx)
+            tables.insert(chunk->GetValue(0, idx).GetValue<std::string>());
+
+        chunk = result->Fetch();
+    }
+
+    std::cout << "\n\n\n\n\nDUCKDB: TABLES SIZE: " << tables.size() << "\n\n\n\n\n";
+
+    return tables;
+}
+
+
+bool DatabaseDuckDB::checkDuckDBTable(const String & table_name) const
+{
+    if (!duckdb_instance)
+        duckdb_instance = openDuckDB(database_path, getContext(), /* throw_on_error */true);
+
+    const String query = fmt::format("SELECT table_name FROM duckdb_tables WHERE table_name='{}';", table_name);
```

Comment:
Can we get arbitrary user string in `table_name` or it's validated somewhere? We can try to use `PreparedStatement` from https://duckdb.org/docs/api/cpp here